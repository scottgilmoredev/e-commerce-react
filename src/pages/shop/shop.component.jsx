import React, { useEffect, lazy, Suspense } from 'react';
import { connect } from 'react-redux';
import { Route } from 'react-router-dom';

// Components
import Spinner from '../../components/spinner/spinner.component';

// Redux actions
import { fetchCollectionsStart } from '../../redux/shop/shop.actions';

// Styles
import { ShopPageContainer } from './shop.styles';

// Lazy loaded components
const CollectionPageContainer = lazy(() => import('../collection/collection.container'));
const CollectionsOverviewContainer = lazy(() =>
    import('../../components/collections-overview/collections-overview.container')
);

/**
 * Display the shop page.
 * This component renders the CollectionOverview and CollectionPage components.
 */
export const ShopPage = ({ fetchCollectionsStart, match }) => {
    useEffect(() => {
        fetchCollectionsStart();
    }, [fetchCollectionsStart]);

    return (
        <ShopPageContainer>
            <Suspense fallback={ <Spinner /> }>
                { /* Collections Overview */ }
                <Route
                    exact
                    path={ `${ match.path }` }
                    component={ CollectionsOverviewContainer }
                />

                { /* Collection Item Page */ }
                <Route
                    path={ `${ match.path }/:collectionId` }
                    component={ CollectionPageContainer }
                />
            </Suspense>
        </ShopPageContainer>
    )
};

const mapDispatchToProps = dispatch => ({
    fetchCollectionsStart: () => dispatch(fetchCollectionsStart()),
});

export default connect(null, mapDispatchToProps)(ShopPage);
