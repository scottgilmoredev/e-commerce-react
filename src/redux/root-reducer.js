import { combineReducers } from 'redux';
import { persistReducer } from 'redux-persist';

// Tell redux-persist to use local storage as default storage.
import storage from 'redux-persist/lib/storage';

// Reducers
import cartReducer from './cart/cart.reducer';
import directoryReducer from './directory/directory.reducer';
import navReducer from './nav/nav.reducer';
import shopReducer from './shop/shop.reducer';
import userReducer from './user/user.reducer';

const persistConfig = {
    key: 'root',
    storage,
    whitelist: ['cart'],
};

const rootReducer = combineReducers({
    cart: cartReducer,
    directory: directoryReducer,
    nav: navReducer,
    shop: shopReducer,
    user: userReducer,
});

export default persistReducer(persistConfig, rootReducer);
