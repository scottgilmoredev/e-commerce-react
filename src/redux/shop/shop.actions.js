import ShopActionTypes from './shop.types';
import { convertCollectionsSnapshotToMap, firestore } from '../../firebase/firebase.utils';

export const fetchCollectionsFailure = errorMessage => ({
    payload: errorMessage,
    type: ShopActionTypes.FETCH_COLLECTIONS_FAILURE,
});

export const fetchCollectionsStart = () => ({
    type: ShopActionTypes.FETCH_COLLECTIONS_START,
});

export const fetchCollectionsSuccess = collectionsMap => ({
    payload: collectionsMap,
    type: ShopActionTypes.FETCH_COLLECTIONS_SUCCESS,
});

/**
 * NOTE: This is to demonstrate testing async functions.
 */
export const fetchCollectionsStartAsync = () => {
    return dispatch => {
        const collectionRef = firestore.collection('collections');
        dispatch(fetchCollectionsStart());

        collectionRef
            .get()
            .then(snapshot => {
                const collectionsMap = convertCollectionsSnapshotToMap(snapshot);
                dispatch(fetchCollectionsSuccess(collectionsMap));
            })
            .catch(error => dispatch(fetchCollectionsFailure(error.message)));
    };
};
