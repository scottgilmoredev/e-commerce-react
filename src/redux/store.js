import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { persistStore } from 'redux-persist';
import createSagaMiddleware from 'redux-saga';

import rootReducer from './root-reducer';
import rootSaga from './root-saga';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) => {
    const middleware = getDefaultMiddleware({
      thunk: false,
      serializableCheck: false,
    }).concat(sagaMiddleware);

    if (process.env.NODE_ENV !== 'production') {
      middleware.push(logger);
    }

    return middleware;
  },
});

sagaMiddleware.run(rootSaga);
export const persistor = persistStore(store);
