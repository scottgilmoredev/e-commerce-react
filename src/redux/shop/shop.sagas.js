import { all, call, takeLatest, put } from 'redux-saga/effects';

// Firebase utils
import { convertCollectionsSnapshotToMap, firestore } from '../../firebase/firebase.utils';

// Redux actions
import { fetchCollectionsFailure, fetchCollectionsSuccess } from './shop.actions';
import ShopActionTypes from './shop.types';

export function* fetchCollectionsAsync() {
    try {
        const collectionRef = firestore.collection('collections');
        const snapshot = yield collectionRef.get();
        const collectionsMap = yield call(convertCollectionsSnapshotToMap, snapshot);

        yield put(fetchCollectionsSuccess(collectionsMap));
    } catch (error) {
        yield put(fetchCollectionsFailure(error.message));
    }
};

export function* onFetchCollectionStart() {
    yield takeLatest(ShopActionTypes.FETCH_COLLECTIONS_START, fetchCollectionsAsync);
};

export function* shopSagas() {
    yield all([
        call(onFetchCollectionStart),
    ]);
};
