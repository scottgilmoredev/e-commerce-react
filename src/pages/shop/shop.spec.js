import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { combineReducers, createStore } from 'redux';
import { Provider } from 'react-redux';

// Component
import { ShopPage } from './shop.component';

export const createMockStore = ({ state, reducers }) => {
    const store = createStore(combineReducers(reducers), state);

    return {
        ...store,
        persistor: {
            persist: () => null,
        },
    };
};

describe('ShopPage', () => {
    let wrapper;
    let mockFetchCollectionsStart;
    let store;

    beforeEach(() => {
        const mockReducer = (
            state = {
                isFetching: true,
            },
            action,
        ) => state;

        const mockState = {
            shop: {
                isFetching: true,
            },
        };

        mockFetchCollectionsStart = jest.fn();

        store = createMockStore({
            state: mockState,
            reducers: { shop: mockReducer },
        });

        const mockMatch = {
            path: '',
        };

        const mockProps = {
            match: mockMatch,
            fetchCollectionsStart: mockFetchCollectionsStart
        };

        wrapper = mount(
            <BrowserRouter>
                <Provider store={ store }>
                    <ShopPage { ...mockProps } />
                </Provider>
            </BrowserRouter>
        );
    });

    // Render the component.
    it('should render the ShopPage component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    // Fetch collections.
    it('should fetch collections', () => {
        expect(mockFetchCollectionsStart).toHaveBeenCalled();
    });
});
