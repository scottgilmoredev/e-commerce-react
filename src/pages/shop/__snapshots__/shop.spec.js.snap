// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopPage should render the ShopPage component 1`] = `ReactWrapper {}`;
